### ISSUE TITLE
Feature: Batch read table data.

### ISSUE THREAD

**Issue by zebin-code** [lancedb/lancedb#1927]
### SDK

Python

### Description

In Lancedb, I need to update the data in a specific column of a table. I have chosen to convert all the from the database into a pandas DataFrame, then iterate through the DataFrame to modify the internal values, and finally use an SQL statement to update the data. The drawback here is that, in the Python version, it seems that batch reading of table data is not supported, and only full reading is allowed, which can lead to excessive memory usage. Is there any expectation for future support in this regard?

**Comment by zebin-code** [comment]
This pattern is similar to when I used the iterator pattern while reading a list, like the one below.

**Comment by zebin-code** [comment]
![impicture_20241209_172504](https://github.com/user-attachments/assets/10147d35-4c7e-4097-ab5a-30c4d845f3c7)

**Comment by wjones127** [comment]
You should be able to scan data in batches just by doing:

```python
for batch in table.query().to_batches(batch_size):
    batch = batch.to_pandas()
    ...
```

**Comment by zebin-code** [comment]
Will the entire table's data be loaded into memory during batch reading here?

**Comment by wjones127** [comment]
> Will the entire table's data be loaded into memory during batch reading here?

No, though LanceDB doesn't have as strong controls over memory use as Lance does. You might want to consider using `.to_lance()` and uses Lance's scanner, which has options like `batch_readahead`, `fragment_readahead`, and `io_buffer_size` to give more control over memory use when scanning.

https://lancedb.github.io/lance/api/python/lance.html#lance.dataset.LanceDataset.scanner

**Comment by zebin-code** [comment]
Thank you for your response. Here, I would like to share my understanding and ask if I am mistaken. During the process of iterating with batches, even in a full table scan, the memory usage remains manageable, as the full table scan utilizes something like an index. Then, when using interfaces like_pandas or to_arrow, the data corresponding to the index is loaded into memory.
![impicture_20241211_111353](https://github.com/user-attachments/assets/f5e94177-5f97-4fae-99e6-b4820196f65f)


Additionally, when using the following statement, does it also involve loading the entire table's data into memory, such as in the statement: [table.search](http://table.search/)().where("select * from table prefilter=True).limit(10)? During the process, will there be a loading of the entire table's data, or will only 10 records be loaded throughout entire process?

**Comment by wjones127** [comment]
> During the process of iterating with batches, even in a full table scan, the memory usage remains manageable, as the full table scan utilizes something like an index.

Yes, the scanner reads one batch at a time. It yields those in a stream / iterator.

> Then, when using interfaces like_pandas or to_arrow, the data corresponding to the index is loaded into memory.

You can think of `to_arrow` or `to_pandas` as calling the scanner, and appending all the batches together until it is all in memory.

> additionally, when using the following statement, does it also involve loading the entire table's data into memory, such as in the statement: [table.search](http://table.search/)().where("select * from table prefilter=True).limit(10)? During the process, will there be a loading of the entire table's data, or will only 10 records be loaded throughout entire process?

The exact process depends on the query and which indexes the table has. You can use `.explain_plan()` to view a query plan. This can tell you whether a scalar index is being used to run the filter, for example.

We use DataFusion to execute our queries. So the execution is similar to other modern OLAP systems. In general, only the data required is loaded from disk, and queries are only executed in a streaming fashion when possible.

**Comment by zebin-code** [comment]
Thank you for your answer, I'll give it a try.
